File: soln-graph-problems/course-schedule-ii-with-indegree.java
/**
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/course-schedule-ii/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    static int[] indegree;
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        ArrayList<Integer> course_order=new ArrayList<>();
        int n=numCourses;
        ArrayList<Integer>[] adj_list=new ArrayList[n];
        for(int i=0;i<n;i++)
            adj_list[i]=new ArrayList<>();
        for(int edge[]:prerequisites)
            adj_list[edge[1]].add(edge[0]);
        
        indegree=new int[n];
        
        for(ArrayList<Integer> nodes:adj_list){
            for(int v:nodes)
                indegree[v]++;
        }
        
        Queue<Integer> queue=new LinkedList<>();
        for(int i=0;i<n;i++){
            if(indegree[i]==0)
                queue.add(i);
        }
        
        int order=0;
        while(!queue.isEmpty()){
            int node=queue.remove();
            course_order.add(node);
            order++;
            for(int adj_node:adj_list[node]){
                indegree[adj_node]--;
                if(indegree[adj_node]==0)
                    queue.add(adj_node);
            }
        }
        
        if(order!=n)
            return new int[0];
        
        int[] result=new int[course_order.size()];
        for(int i=0;i<course_order.size();i++){
            result[i]=course_order.get(i);
        }
        
        return result;
    }
}

File: C++/soln-graph-problems/course-schedule-ii-with-indegree.cpp
/**
 @author Farheen Bano
  
 Date 24-07-2021

 Reference-
 https://leetcode.com/problems/course-schedule-ii/
*/

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> adj_list(numCourses);
        vector<int> course_order;   
        vector<int> indegree(numCourses,0);   

        for(auto edge:prerequisites)
            adj_list[edge[1]].push_back(edge[0]);
        
        for(auto nodes:adj_list){
            for(int v:nodes)
                indegree[v]++;
        }
        
        queue<int> q;
        for(int i=0;i<numCourses;i++){
            if(indegree[i]==0)
                q.push(i);
        }
        
        int takenCourses=0;
        while(!q.empty()){
            int node=q.front();
            q.pop();
            course_order.push_back(node);
            takenCourses++;
            for(int adj_node:adj_list[node]){
                indegree[adj_node]--;
                if(indegree[adj_node]==0)
                    q.push(adj_node);
            }
        }
        
        vector<int> res;
        if(takenCourses!=numCourses)
            return res;
        vector<int> result(course_order.size());
        for(int i=0;i<course_order.size();i++){
            result[i]=course_order[i];
        }
        return result;
    }
};
